src/coreclr/pal/src/thread/thread.cpp

```diff
@@ -74,7 +69,7 @@ SET_DEFAULT_DEBUG_CHANNEL(THREAD); // some headers have code with asserts, so do
 #include <lwp.h>
 #endif
 // If we don't have sys/lwp.h but do expect to use _lwp_self, declare it to silence compiler warnings
-#if HAVE__LWP_SELF && !HAVE_SYS_LWP_H && !HAVE_LWP_H
+#if HAVE__LWP_SELF && !defined(HAVE_SYS_LWP_H) && !defined(HAVE_LWP_H)
 extern "C" int _lwp_self ();
 #endif
```

Comment:
Note: `HAVE_SYS_LWP_H` and `HAVE_LWP_H` are defined here, but as zero.
Now that `_KERNEL` is not defined in here, the _lwp_self() declaration is visible,
so we could probably just get rid of the three lines above.

Reply:
Do any systems other than SunOS derived ones have `lwp.h` or `sys/lwp.h`?

Reply:
This is defined as:

```
src/coreclr/pal/src/config.h.in:#cmakedefine01 HAVE_SYS_LWP_H
src/coreclr/pal/src/configure.cmake:check_include_files(sys/lwp.h HAVE_SYS_LWP_H)
```
so the header is rendered with `#define HAVE_SYS_LWP_H 0` or 1 depending on the introspection results (cmakedefine01 vs. cmakedefine). LWP concepts are shared with NetBSD.

Here's a summary of {sys/}lwp.h:

    NetBSD: Yes
    FreeBSD: No
    Linux: No
    macOS: No
    Illumos: Yes

Others platform (with No) have their own threading models.

Option 1: add `!defined(__sun)` or delete . 

Option 2: we can clean these up while at it:

```sh
$ git grep -E 'HAVE_LWP_H|HAVE_SYS_LWP_H' 
src/coreclr/pal/src/config.h.in:#cmakedefine01 HAVE_SYS_LWP_H
src/coreclr/pal/src/config.h.in:#cmakedefine01 HAVE_LWP_H
src/coreclr/pal/src/configure.cmake:check_include_files(sys/lwp.h HAVE_SYS_LWP_H)
src/coreclr/pal/src/configure.cmake:check_include_files(lwp.h HAVE_LWP_H)
src/coreclr/pal/src/thread/thread.cpp:#if HAVE_SYS_LWP_H
src/coreclr/pal/src/thread/thread.cpp:#if HAVE_LWP_H
src/coreclr/pal/src/thread/thread.cpp:#if HAVE__LWP_SELF && !HAVE_SYS_LWP_H && !HAVE_LWP_H
```

I'd go for option 2 in this case since I haven't tested build on NetBSD in ages, not sure if anyone else has either.

Reply:
That's interesting. I put some temporary sections in the code to check on this, eg:
```
#ifdef __sun
#if !defined(HAVE_SYS_LWP_H)
#error "notdef HAVE_SYS_LWP_H"
#endif
#endif
```
etc.  (I've dropped the details, but I could recreate if you need it)
That's how I figured out those were defined but we saw the extra declaration anyway.

Reply:
It is documented here: https://cmake.org/cmake/help/latest/command/configure_file.html.

```
#cmakedefine01 VAR

will be replaced with either

#define VAR 0

or

#define VAR 1
```

If cmakedefine01 is not working as documented, then it will break a lot of code in this repo.

Reply:
I've verified that NetBSD `lwp.h` provides a declaration for `_lwp_sef()`.
I'm now confident that the local declaration was added here for sunos, and was needed because the `_KERNEL` define masked that declaration.   With that fixed, this local declaration (hack) is no longer needed.  Let's just remove it.   I updated this PR to do that.

Reply:
OK, these were their only usages, so please also delete `git grep -E 'HAVE_SYS_LWP_H|HAVE_SYS_LWP_H|HAVE_LWP_H'`.

Reply:
All three of these definitions are still used:
```
git ls-files -z |xargs -0 egrep -l '(HAVE__LWP_SELF|HAVE_LWP_H|HAVE_SYS_LWP_H)'
src/coreclr/pal/src/config.h.in
src/coreclr/pal/src/configure.cmake
src/coreclr/pal/src/thread/thread.cpp
```
I looked over the usages in `thread.cpp` and they look sensible to me.

Reply:
I thought the only usage was related to the line you have just removed. Are `#include <lwp.h>` etc. still needed?

Reply:
Yes, they are still used and needed.  See `thread.cpp`

Reply:
This is thread.cpp we are commenting on.
The only remaining usage is #include. So you are saying we are using something from that header, header is under #if HAVE_XX condition and that usage is not?

Reply:
This is thread.cpp we are commenting on.
The only remaining usage of HAVE_LWP_H and HAVE_SYS_LWP_H is to condition `#include <{sys/}lwp.h>`. So you are saying we are using something from that header, header is included under #if HAVE_XX condition and that usage is not?

Reply:
Here are the remaining usages I'm referring to.  I think these should stay:

thread.cpp : 65
```
#if HAVE_SYS_LWP_H
#include <sys/lwp.h>
#endif
#if HAVE_LWP_H
#include <lwp.h>
#endif
```
thread.cpp : 645
```
#if HAVE_THREAD_SELF || HAVE__LWP_SELF
    /* Create new threads as "bound", so each pthread is permanently bound
       to an LWP.  Get/SetThreadContext() depend on this 1:1 mapping. */
    pthread_attr_setscope(&pthreadAttr, PTHREAD_SCOPE_SYSTEM);
#endif // HAVE_THREAD_SELF || HAVE__LWP_SELF
```
thread.cpp : 1683
```
#if HAVE_THREAD_SELF
    pThread->m_dwLwpId = (DWORD) thread_self();
#elif HAVE__LWP_SELF
    pThread->m_dwLwpId = (DWORD) _lwp_self();
#else
    pThread->m_dwLwpId = 0;
#endif
```

The first gets the declaration of `_lwp_self()`.
The second and third do stuff dependent on `HAVE__LWP_SELF`

I don't think any of that is affected by the proposed change.
Am I'm missing something?

Thanks.